File: bank-service/src/main/java/rs/edu/raf/controller/CreditController.java
package rs.edu.raf.controller;

import io.swagger.annotations.ApiOperation;
import io.swagger.v3.oas.annotations.Parameter;
import io.swagger.v3.oas.annotations.security.SecurityRequirement;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import lombok.AllArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import rs.edu.raf.dto.CreditDTO;
import rs.edu.raf.dto.CreditRequestCreateDto;
import rs.edu.raf.dto.CreditRequestDTO;
import rs.edu.raf.security.JwtUtil;
import rs.edu.raf.service.CreditService;

import java.util.List;

@RestController
@RequestMapping("/credit")
@AllArgsConstructor
@Tag(name = "Credits", description = "Managing credits")
@SecurityRequirement(name = "jwt")
public class CreditController {
    private final CreditService creditService;
    private final JwtUtil jwtUtil;

    @ApiOperation(value = "Apply for credit")
    @PostMapping("/apply")
    public ResponseEntity<CreditRequestDTO> applyForCredit(@RequestBody @Valid CreditRequestCreateDto creditRequestCreateDto){
        return new ResponseEntity<>(creditService.applyForCredit(creditRequestCreateDto), HttpStatus.OK);
    }

    @ApiOperation(value = "Deny credit request")
    @PutMapping("/deny/{id}")
    public ResponseEntity<CreditRequestDTO> denyCreditRequest(@Parameter(name = "Credit request ID")
                                                                  @PathVariable("id") Long id) {
        return new ResponseEntity<>(creditService.dennyCreditRequest(id), HttpStatus.OK);
    }

    @ApiOperation(value = "Approve credit request")
    @PutMapping("/approve/{id}")
    public ResponseEntity<CreditRequestDTO> approveCredit(@Parameter(name = "Credit request ID") @PathVariable("id") Long id){
        return new ResponseEntity<>(creditService.approveCreditRequest(id), HttpStatus.OK);
    }

    @ApiOperation(value = "Get all credit requests with specified status")
    @GetMapping("/requests/all/{status}")
    public ResponseEntity<List<CreditRequestDTO>> getAllCreditRequestsWithStatus(@Parameter(name = "Credit request status")
                                                                                    @PathVariable("status") String status){
        return new ResponseEntity<>(creditService.getAllCreditRequestsWithStatus(status), HttpStatus.OK);
    }

    @ApiOperation(value = "Get all credit requests with specified status for user")
    @GetMapping("/requests/{status}")
    public ResponseEntity<List<CreditRequestDTO>> getAllCreditRequestsForUserWithStatus(@Parameter(name = "Credit request status") @PathVariable("status") String status) {
        return new ResponseEntity<>(creditService.getAllCreditRequestForUserWithStatus(jwtUtil.getIDForLoggedUser(), status), HttpStatus.OK);
    }

    @ApiOperation("Get all credits")
    @GetMapping("/all")
    public ResponseEntity<List<CreditDTO>> getAllCredits() {
        return new ResponseEntity<>(creditService.getAllCredits(), HttpStatus.OK);
    }

    @ApiOperation("Get all credits")
    @GetMapping
    public ResponseEntity<List<CreditDTO>> getAllCreditsForUser() {
        return new ResponseEntity<>(creditService.getAllCredits(), HttpStatus.OK);
    }

    @ApiOperation("Get details about credit")
    @GetMapping("/{id}")
    public ResponseEntity<CreditDTO> getCreditDetails(@Parameter(name = "Credit ID") @PathVariable("id") Long id){
        return new ResponseEntity<>(creditService.getDetailedCredit(id), HttpStatus.OK);
    }

}

File: bank-service-kotlin/src/main/kotlin/org/example/bankservicekotlin/controller/CreditController.kt
package org.example.bankservicekotlin.controller

import jakarta.validation.Valid
import org.example.bankservicekotlin.dto.CreditDTO
import org.example.bankservicekotlin.dto.CreditRequestCreateDTO
import org.example.bankservicekotlin.dto.CreditRequestDTO
import org.example.bankservicekotlin.security.JwtUtil
import org.example.bankservicekotlin.service.CreditService
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/credit")
class CreditController(
    private val creditService: CreditService, private val jwtUtil: JwtUtil
) {
    @PostMapping("/apply")
    fun applyForCredit(@RequestBody creditRequestCreateDto: @Valid CreditRequestCreateDTO): ResponseEntity<CreditRequestDTO> {
        return ResponseEntity<CreditRequestDTO>(creditService.applyForCredit(creditRequestCreateDto), HttpStatus.OK)
    }

    @PutMapping("/deny/{id}")
    fun denyCreditRequest(@PathVariable("id") id: Long): ResponseEntity<CreditRequestDTO> {
        return ResponseEntity<CreditRequestDTO>(creditService.dennyCreditRequest(id), HttpStatus.OK)
    }

    @PutMapping("/approve/{id}")
    fun approveCredit(@PathVariable("id") id: Long): ResponseEntity<CreditRequestDTO> {
        return ResponseEntity<CreditRequestDTO>(creditService.approveCreditRequest(id), HttpStatus.OK)
    }

    @GetMapping("/requests/all/{status}")
    fun getAllCreditRequestsWithStatus(@PathVariable("status") status: String): ResponseEntity<List<CreditRequestDTO>> {
        return ResponseEntity<List<CreditRequestDTO>>(
            creditService.getAllCreditRequestsWithStatus(status),
            HttpStatus.OK
        )
    }

    @GetMapping("/requests/{status}")
    fun getAllCreditRequestsForUserWithStatus(@PathVariable("status") status: String): ResponseEntity<List<CreditRequestDTO>> {
        return ResponseEntity<List<CreditRequestDTO>>(
            creditService.getAllCreditRequestForUserWithStatus(
                jwtUtil.getIDForLoggedUser(),
                status
            ), HttpStatus.OK
        )
    }

    @GetMapping("/all")
    fun getAllCredits(): ResponseEntity<List<CreditDTO>> {
        return ResponseEntity<List<CreditDTO>>(creditService.getAllCredits(), HttpStatus.OK)
    }

    @GetMapping
    fun getAllCreditsForUser(): ResponseEntity<List<CreditDTO>> {
        return ResponseEntity<List<CreditDTO>>(creditService.getAllCredits(), HttpStatus.OK)
    }

    @GetMapping("/{id}")
    fun getCreditDetails(@PathVariable("id") id: Long): ResponseEntity<CreditDTO> {
        return ResponseEntity<CreditDTO>(creditService.getDetailedCredit(id), HttpStatus.OK)
    }
}